mypy/fastparse.py

```diff
@@ -0,0 +1,744 @@
+    FloatExpr, CallExpr, SuperExpr, MemberExpr, IndexExpr, SliceExpr, OpExpr,
+    UnaryExpr, FuncExpr, ComparisonExpr,
+    StarExpr, YieldFromExpr, NonlocalDecl, DictionaryComprehension,
+    SetComprehension, ComplexExpr, EllipsisExpr, YieldExpr, Argument,
+    ARG_POS, ARG_OPT, ARG_STAR, ARG_NAMED, ARG_STAR2
+)
+from mypy.types import Type, CallableType, AnyType, UnboundType, TupleType, TypeList, EllipsisType
+from mypy import defaults
+from mypy.errors import Errors
+
+try:
+    import typed_ast  # type: ignore
+except ImportError:
+    print('You must install the typed_ast module before you can run mypy with `--fast-parser`.\n'
+          'The typed_ast module can be found at https://github.com/ddfisher/typed_ast',
+          file=sys.stderr)
+    sys.exit(1)
+
+T = TypeVar('T')
+
+
+def parse(source: Union[str, bytes], fnam: str = None, errors: Errors = None,
+          pyversion: Tuple[int, int] = defaults.PYTHON3_VERSION,
+          custom_typing_module: str = None, implicit_any: bool = False) -> MypyFile:
+    """Parse a source file, without doing any semantic analysis.
+
+    Return the parse tree. If errors is not provided, raise ParseError
+    on failure. Otherwise, use the errors object to report parse errors.
+
+    The pyversion (major, minor) argument determines the Python syntax variant.
+    """
+    is_stub_file = bool(fnam) and fnam.endswith('.pyi')
+    try:
+        ast = typed_ast.parse(source, fnam, 'exec')
+    except SyntaxError as e:
+        if errors:
+            errors.set_file('<input>' if fnam is None else fnam)
+            errors.report(e.lineno, e.msg)  # type: ignore
+        else:
+            raise
+    else:
+        tree = ASTConverter().visit(ast)
+        tree.path = fnam
+        tree.is_stub = is_stub_file
+        return tree
+
+    return MypyFile([],
+                    [],
+                    False,
+                    set(),
+                    weak_opts=set())
+
+
+def parse_type_comment(type_comment: str, line: int) -> Type:
+    typ = typed_ast.parse(type_comment, '<type_comment>', 'eval')
+    return TypeConverter(line=line).visit(typ.body)
+
+
+def with_line(f):
+    @wraps(f)
+    def wrapper(self, ast):
+        node = f(self, ast)
+        node.set_line(ast.lineno)
+        return node
+    return wrapper
+
+
+def find(f: Callable[[T], bool], seq: Sequence[T]) -> T:
+    for item in seq:
+        if f(item):
+            return item
+    return None
+
+
+class ASTConverter(typed_ast.NodeTransformer):
+    def __init__(self):
+        self.in_class = False
+        self.imports = []
+
+    def generic_visit(self, node):
+        raise RuntimeError('AST node not implemented: ' + str(type(node)))
+
+    def visit_NoneType(self, n):
+        return None
+
+    def visit_list(self, l):
+        return [self.visit(e) for e in l]
+
+    def from_operator(self, op):
+        if isinstance(op, typed_ast.Add):
+            return '+'
```

Comment:
Hmm.  And doing a lookup on `type(op)`?  Yeah, I think that should work.  I'll try it out.